File: go/algorithms/odd_even.go
package algorithms

// OddEvenSort sorts in-place by alternating odd and even phases.
func OddEvenSort(arr []int) {
    n := len(arr)
    sorted := false
    for !sorted {
        sorted = true
        for i := 1; i <= n-2; i += 2 {
            if arr[i] > arr[i+1] {
                arr[i], arr[i+1] = arr[i+1], arr[i]
                sorted = false
            }
        }
        for i := 0; i <= n-2; i += 2 {
            if arr[i] > arr[i+1] {
                arr[i], arr[i+1] = arr[i+1], arr[i]
                sorted = false
            }
        }
    }
}

File: rust/src/odd_even.rs
pub fn odd_even_sort(arr: &mut [i32]) {
    let mut sorted = false;
    let n = arr.len();
    while !sorted {
        sorted = true;
        let mut i = 1;
        while i + 1 < n {
            if arr[i] > arr[i + 1] {
                arr.swap(i, i + 1);
                sorted = false;
            }
            i += 2;
        }
        i = 0;
        while i + 1 < n {
            if arr[i] > arr[i + 1] {
                arr.swap(i, i + 1);
                sorted = false;
            }
            i += 2;
        }
    }
}
